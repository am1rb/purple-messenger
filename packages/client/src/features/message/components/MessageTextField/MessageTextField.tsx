import React, { useRef, useCallback } from "react";
import { debounce } from "throttle-debounce";
import UnformTextField, {
  Props as UnformTextFieldProps,
} from "components/UnformTextField";

interface Props extends Omit<UnformTextFieldProps, "multiline" | "fullWidth"> {
  onTypeStart?: () => void;
  onTypeStop?: () => void;
  onEnter?: () => void;
  detectTypeTime?: number;
}

function MessageTextField({
  onTypeStart,
  onTypeStop,
  detectTypeTime = 1500,
  onEnter,
  onChange,
  onKeyDown,
  ...other
}: Props) {
  const isTyping = useRef(false);
  const handleStartTyping = useCallback(
    debounce(detectTypeTime, true, () => {
      isTyping.current = true;
      onTypeStart?.();
    }),
    [detectTypeTime, onTypeStart]
  );
  const handleStopTyping = useCallback(
    debounce(detectTypeTime, () => {
      if (isTyping.current) {
        isTyping.current = false;
        onTypeStop?.();
      }
    }),
    [detectTypeTime, onTypeStop]
  );

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      handleStartTyping();
      handleStopTyping();
      onChange?.(event);
    },
    [handleStartTyping, handleStopTyping, onChange]
  );

  const handleKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>) => {
      if (event.keyCode === 13 && !event.shiftKey) {
        isTyping.current = false;
        onTypeStop?.();
        onEnter?.();
        event.preventDefault();
      }
      onKeyDown?.(event);
    },
    [onEnter, onKeyDown]
  );

  return (
    <UnformTextField
      {...other}
      multiline
      fullWidth
      onChange={handleChange}
      onKeyDown={handleKeyDown}
    />
  );
}

export default MessageTextField;
