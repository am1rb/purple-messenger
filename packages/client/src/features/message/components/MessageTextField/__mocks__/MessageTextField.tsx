import React from "react";
import useTextFieldEvents, {
  TextFieldEventsProps,
} from "../../useTextFieldEvents";

type InputProps = React.HTMLAttributes<HTMLInputElement>;
export type MessageTextFieldProps = Omit<InputProps, "onChange" | "onKeyDown"> &
  TextFieldEventsProps<InputProps>;

function MessageTextField({
  onStartTyping,
  onStopTyping,
  detectTypeTime,
  onEnter,
  onChange,
  onKeyDown,
  ...other
}: MessageTextFieldProps) {
  const { handleChange, handleKeyDown } = useTextFieldEvents<InputProps>({
    onStartTyping,
    onStopTyping,
    detectTypeTime,
    onEnter,
    onChange,
    onKeyDown,
  });

  return (
    <input
      {...other}
      data-testid="mock-message-text-field"
      onChange={handleChange}
      onKeyDown={handleKeyDown}
    />
  );
}

export default MessageTextField;
