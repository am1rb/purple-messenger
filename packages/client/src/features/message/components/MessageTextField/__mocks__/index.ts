export { default } from "./MessageTextField";
export * from "./MessageTextField";
