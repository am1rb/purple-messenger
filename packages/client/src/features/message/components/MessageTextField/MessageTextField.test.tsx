import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import MessageTextField, { MessageTextFieldProps } from "./MessageTextField";

jest.mock("components/UnformTextField");

const sharedProps: MessageTextFieldProps = {
  name: "field-name",
  detectTypeTime: 200,
};

describe("The <UnformTextField /> tests", () => {
  it("Should match the snapshot", () => {
    const { container } = render(<MessageTextField {...sharedProps} />);
    expect(container).toMatchSnapshot();
  });

  it("Should call onStartTyping properly", () => {
    const handleStartTyping = jest.fn();
    const { getByTestId } = render(
      <MessageTextField {...sharedProps} onStartTyping={handleStartTyping} />
    );
    userEvent.type(getByTestId("mock-unform-text-field"), "Hi");
    expect(handleStartTyping).toHaveBeenCalled();
  });

  it("Should call onStopTyping properly", async () => {
    const handleStopTyping = jest.fn();
    const { getByTestId } = render(
      <MessageTextField {...sharedProps} onStopTyping={handleStopTyping} />
    );
    userEvent.type(getByTestId("mock-unform-text-field"), "Hi");
    await wait(() => expect(handleStopTyping).toHaveBeenCalled());
  });

  it("Should call onEnter properly", () => {
    const handleEnter = jest.fn();
    const { getByTestId } = render(
      <MessageTextField {...sharedProps} onEnter={handleEnter} />
    );

    const input = getByTestId("mock-unform-text-field");

    userEvent.type(input, "Hello");
    fireEvent.keyDown(input, { key: "Enter", keyCode: 13 });

    expect(handleEnter).toHaveBeenCalled();
  });
});
