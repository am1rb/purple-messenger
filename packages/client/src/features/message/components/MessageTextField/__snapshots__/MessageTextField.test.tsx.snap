// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <UnformTextField /> tests Should match the snapshot 1`] = `
<div>
  <textarea
    data-testid="mock-unform-text-field"
    name="field-name"
  />
</div>
`;
